clouddriver-kubernetes/src/test/java/com/netflix/spinnaker/clouddriver/kubernetes/caching/view/provider/KubernetesDataProviderIntegrationTest.java

```diff
@@ -0,0 +1,1117 @@
+      new DefaultProviderRegistry(
+          ImmutableList.of(kubernetesV2Provider), new InMemoryNamedCacheFactory());
+  private static final KubernetesCacheUtils cacheUtils =
+      new KubernetesCacheUtils(
+          providerRegistry.getProviderCache(kubernetesV2Provider.getProviderName()),
+          kindMap,
+          accountResolver);
+  private static final ImmutableSetMultimap<String, String> manifestsByNamespace =
+      ImmutableSetMultimap.<String, String>builder()
+          .putAll(
+              "backend-ns",
+              ImmutableSet.of(
+                  "backend-service.yml",
+                  "backend-rs-014.yml",
+                  "backend-pod-014.yml",
+                  "backend-rs-015.yml",
+                  "backend-pod-015.yml"))
+          .putAll(
+              "frontend-ns",
+              ImmutableSet.of(
+                  "frontend-service.yml",
+                  "frontend-deployment.yml",
+                  "frontend-rs-old.yml",
+                  "frontend-rs-new.yml",
+                  "frontend-pod-1.yml",
+                  "frontend-pod-2.yml"))
+          .build();
+
+  private static KubernetesV2ApplicationProvider applicationProvider =
+      new KubernetesV2ApplicationProvider(cacheUtils);
+  private static KubernetesV2ClusterProvider clusterProvider =
+      new KubernetesV2ClusterProvider(cacheUtils, kindMap);
+  private static KubernetesV2InstanceProvider instanceProvider =
+      new KubernetesV2InstanceProvider(cacheUtils, accountResolver);
+  private static KubernetesV2LoadBalancerProvider loadBalancerProvider =
+      new KubernetesV2LoadBalancerProvider(cacheUtils, kindMap);
+  private static KubernetesV2SearchProvider searchProvider =
+      new KubernetesV2SearchProvider(cacheUtils, kindMap, objectMapper, accountResolver);
+  private static KubernetesV2ServerGroupManagerProvider serverGroupManagerProvider =
+      new KubernetesV2ServerGroupManagerProvider(cacheUtils);
+
+  @BeforeAll
+  static void prepareCache() {
+    KubernetesNamedAccountCredentials<KubernetesV2Credentials> credentials =
+        getNamedAccountCredentials();
+    credentialsRepository.save(credentials.getName(), credentials);
+    dispatcher
+        .buildAllCachingAgents(credentials)
+        .forEach(agent -> agent.getAgentExecution(providerRegistry).executeAgent(agent));
+  }
+
+  @Test
+  void getClusters(SoftAssertions softly) {
+    Map<String, Set<KubernetesV2Cluster>> results = clusterProvider.getClusters();
+    assertThat(results).hasSize(1);
+    assertThat(results).containsKey(ACCOUNT_NAME);
+
+    Set<KubernetesV2Cluster> clusters = results.get(ACCOUNT_NAME);
+    assertThat(clusters).hasSize(4);
+
+    // TODO(ezimanyi): I don't think it's correct that we are returning services/load balancers as
+    // clusters. I believe that they are just being ignored by deck, and should be removed from
+    // the result of this call (likely with performance benefits). In any case, as seen in the
```

Comment:
👍  Agree that only entities mapped to Spinnaker server groups should be returned with clusters. We are even [documenting](https://spinnaker.io/concepts/#cluster) this as such already.

However, some users may expect this to work when using dynamic target selection in, for instance, a Delete or Patch (Manifest) stage, which currently permits selection of any registered kind when choosing a target dynamically. The Spinnaker-cluster concept is a bit more intuitive when configuring Enable/Disable/Scale stages, which work for server group types only, but likely pretty unintuitive for Patch/Delete, so perhaps we can rename the "Choose a target dynamically," which implies that one can choose _any_ target dynamically, to "Choose a target dynamically by cluster" or something, and then only hydrate the resulting Kinds list with Kinds that actually have cluster relationships.

Reply:
Great point!  It looks like [getClusterAndSortAscending](https://github.com/spinnaker/clouddriver/blob/2b78b4693dccce18b87e7fea20dd32100099a2c1/clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/caching/view/provider/KubernetesV2ManifestProvider.java#L116) is expecting a relationship between the cluster and the kind that is being deployed.  We currently only map a few kinds to `cluster` which are (except for `Service` and `NetworkPolicy`) all server-group type things.  So this currently only works for a subset of kinds, and I agree that it probably makes sense to just be clear that this only works for server-group type things.